cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -121,6 +122,168 @@ Result<std::unique_ptr<KernelState>> CountInit(KernelContext*,
+}
+
+template <typename Type, typename VisitorArgType>
+void AddCountDistinctKernel(InputType type, ScalarAggregateFunction* func) {
+  AddAggKernel(KernelSignature::Make({type}, ValueDescr::Scalar(int64())),
+               aggregate::CountDistinctInit<Type, VisitorArgType>, func);
+}
+
+template <typename Type, typename VisitorArgType = typename Type::c_type>
+struct CountDistinctKernel {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    using PhysicalType = typename Type::PhysicalType;
+    AddCountDistinctKernel<PhysicalType, VisitorArgType>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<FixedSizeBinaryType, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<DecimalType, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<Decimal128Type, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<Decimal256Type, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+void AddCountDistinctKernels(ScalarAggregateFunction* func) {
+  // Boolean
+  aggregate::CountDistinctKernel<BooleanType>::Add(boolean(), func);
+  // Number
+  aggregate::CountDistinctKernel<Int8Type>::Add(int8(), func);
+  aggregate::CountDistinctKernel<Int16Type>::Add(int16(), func);
+  aggregate::CountDistinctKernel<Int32Type>::Add(int32(), func);
+  aggregate::CountDistinctKernel<Int64Type>::Add(int64(), func);
+  aggregate::CountDistinctKernel<UInt8Type>::Add(uint8(), func);
+  aggregate::CountDistinctKernel<UInt16Type>::Add(uint16(), func);
+  aggregate::CountDistinctKernel<UInt32Type>::Add(uint32(), func);
+  aggregate::CountDistinctKernel<UInt64Type>::Add(uint64(), func);
+  aggregate::CountDistinctKernel<HalfFloatType>::Add(float16(), func);
+  aggregate::CountDistinctKernel<FloatType>::Add(float32(), func);
+  aggregate::CountDistinctKernel<DoubleType>::Add(float64(), func);
+  // Date
+  aggregate::CountDistinctKernel<Date32Type>::Add(date32(), func);
+  aggregate::CountDistinctKernel<Date64Type>::Add(date64(), func);
+  // Time
+  aggregate::CountDistinctKernel<Time32Type>::Add(time32(TimeUnit::SECOND), func);
+  aggregate::CountDistinctKernel<Time32Type>::Add(time32(TimeUnit::MILLI), func);
+  aggregate::CountDistinctKernel<Time64Type>::Add(time64(TimeUnit::MICRO), func);
+  aggregate::CountDistinctKernel<Time64Type>::Add(time64(TimeUnit::NANO), func);
+  // Timestamp & Duration
```

Comment:
I still think this can be simplified, so I am going to try out something and let you know if it works.

Reply:
Thanks, feel free to share your ideas too!

Reply:
@edponce btw I reduced those calls using the type matchers, I guess this is resolved, let me know otherwise!